File: wgmy23/linux_mem_usage/solve.py
# tried to write it in C++, but the memUsed function just didnt iterate through the child's children for some reason
# but this doesn't work cause of the time limit, so gotta go back to C++ ig
# was it a pointer issue? I didnt use current as a Process* and also didnt use new? idk. prob
# cause before, the Processes were all stored in the same memory address mcm lol

class Process:
    def __init__(self,a,b,c):
        self.pid = a
        self.ppid = b
        self.mem = c
        self.children = []
    
    def memUsed(self):
        used = self.mem
        for child in self.children:
            used += child.memUsed()

        return used


allProcess = {}
n,q = [int(a) for a in input().split()]

for i in range(n):
    a,b,c =  [int(j) for j in input().split()]
    current = Process(a,b,c)
    allProcess[a] = current

    if b != 0:
        allProcess[b].children.append(current)

for i in range(q):
    pid = int(input())
    print(allProcess[pid].memUsed())

File: wgmy23/linux_mem_usage/solve.cpp
#include <bits/stdc++.h>

using namespace std;

class Process {
    public:
        int pid;
        int ppid;
        int mem;
        int totalmem = 0;
        vector<Process*> children;

        Process(){
            pid = 0;
            ppid = 0;
            mem = 0;
        }

        Process(int a, int b, int c){
            pid = a;
            ppid = b;
            mem = c;
        }

        int memUsed(){
            if (totalmem != 0){
                // this has alr been calculated
                return totalmem;
            }

            int total = mem;
            for (Process* child : children){
                total += child->memUsed();
            }
            
            totalmem = total;
            return total;
        }
};

int main(){
    int n,q;
    cin >> n >> q;

    int a,b,c;
    unordered_map <int,Process*> hashmap;

    for (int i = 0; i < n; i++) {
        cin >> a >> b >> c;

        Process *current = new Process(a,b,c);
        hashmap.insert({a,current});

        if (b != 0)
            hashmap[b]->children.push_back(current);  
    }

    int pid;
    for (int i = 0; i < q; i++){
        cin >> pid;
        cout << hashmap[pid]->memUsed() << endl;
    }

    return 0;
}
